File: ios-native/Sources/SportkalenderiOS/ViewModel/SportkalenderViewModel.swift
import Foundation

@MainActor
final class SportkalenderViewModel: ObservableObject {
    @Published var selectedLeague: LeagueOption = leagues.first!
    @Published var selectedSeason: SeasonOption = seasons.first!
    @Published var weekOffset: Int = 0
    @Published var isLoading = false
    @Published var error: String?
    @Published var weekLabel = ""
    @Published var dayBlocks: [DayBlock] = []

    private let service = OpenLigaService()
    private var allMatches: [OpenLigaMatch] = []

    static let leagues: [LeagueOption] = [
        .init(id: "bl1", label: "Bundesliga 1"),
        .init(id: "bl2", label: "Bundesliga 2"),
        .init(id: "bl3", label: "Bundesliga 3"),
        .init(id: "dfb", label: "DFB-Pokal"),
        .init(id: "ucl", label: "Champions League")
    ]

    static let seasons: [SeasonOption] = [
        .init(id: "2025", label: "2025/26"),
        .init(id: "2024", label: "2024/25"),
        .init(id: "2023", label: "2023/24")
    ]

    func onAppear() {
        if dayBlocks.isEmpty { Task { await reload() } }
    }

    func reload() async {
        isLoading = true
        error = nil
        do {
            allMatches = try await service.loadMatches(league: selectedLeague.id, season: selectedSeason.id)
            isLoading = false
            recalcWeek()
        } catch {
            isLoading = false
            self.error = error.localizedDescription
        }
    }

    func changeLeague(_ league: LeagueOption) {
        selectedLeague = league
        weekOffset = 0
        Task { await reload() }
    }

    func changeSeason(_ season: SeasonOption) {
        selectedSeason = season
        weekOffset = 0
        Task { await reload() }
    }

    func previousWeek() {
        weekOffset -= 1
        recalcWeek()
    }

    func nextWeek() {
        weekOffset += 1
        recalcWeek()
    }

    func currentWeek() {
        weekOffset = 0
        recalcWeek()
    }

    private func recalcWeek() {
        let cal = Calendar(identifier: .gregorian)
        let tz = TimeZone(identifier: "Europe/Berlin")!
        var comps = cal.dateComponents(in: tz, from: Date())
        comps.hour = 0; comps.minute = 0; comps.second = 0
        let today = cal.date(from: comps) ?? Date()
        let weekday = cal.component(.weekday, from: today)
        let mondayOffset = weekday == 1 ? -6 : (2 - weekday)
        let monday = cal.date(byAdding: .day, value: mondayOffset + (weekOffset * 7), to: today) ?? today
        let sunday = cal.date(byAdding: .day, value: 6, to: monday) ?? monday

        let labelFmt = DateFormatter()
        labelFmt.timeZone = tz
        labelFmt.locale = Locale(identifier: "de_DE")
        labelFmt.dateFormat = "dd.MM.yyyy"

        weekLabel = "\(labelFmt.string(from: monday)) - \(labelFmt.string(from: sunday))"

        let dayFmt = DateFormatter()
        dayFmt.timeZone = tz
        dayFmt.locale = Locale(identifier: "de_DE")
        dayFmt.dateFormat = "EEEE"

        let timeFmt = DateFormatter()
        timeFmt.timeZone = tz
        timeFmt.locale = Locale(identifier: "de_DE")
        timeFmt.dateFormat = "HH:mm"

        dayBlocks = (0...6).compactMap { d in
            guard let date = cal.date(byAdding: .day, value: d, to: monday) else { return nil }
            let dateLabel = labelFmt.string(from: date)
            let matches = allMatches
                .filter { m in
                    guard let kickoff = m.kickoffDate() else { return false }
                    return cal.isDate(kickoff, inSameDayAs: date)
                }
                .sorted { ($0.kickoffDate() ?? .distantFuture) < ($1.kickoffDate() ?? .distantFuture) }
                .map { m -> MatchItem in
                    let kickoff = m.kickoffDate()
                    let score: String = finalScore(m) ?? "vs"
                    let id = "\(dateLabel)-\(m.team1?.teamName ?? "x")-\(m.team2?.teamName ?? "y")-\(timeFmt.string(from: kickoff ?? date))"
                    return MatchItem(
                        id: id,
                        kickoff: kickoff.map { timeFmt.string(from: $0) } ?? "--:--",
                        home: m.team1?.teamName ?? "Unbekannt",
                        away: m.team2?.teamName ?? "Unbekannt",
                        score: score,
                        tvChannels: resolveTvChannels(league: selectedLeague.id, kickoff: kickoff)
                    )
                }

            return DayBlock(
                id: dateLabel,
                dayLabel: dayFmt.string(from: date).capitalized,
                dateLabel: dateLabel,
                isToday: cal.isDate(date, inSameDayAs: today),
                matches: matches
            )
        }
    }

    private func finalScore(_ match: OpenLigaMatch) -> String? {
        let final = match.matchResults?
            .sorted(by: { ($0.resultTypeID ?? 0) > ($1.resultTypeID ?? 0) })
            .first(where: { $0.pointsTeam1 != nil && $0.pointsTeam2 != nil })

        if let f = final, let a = f.pointsTeam1, let b = f.pointsTeam2 {
            return "\(a):\(b)"
        }
        return nil
    }
}

let leagues = SportkalenderViewModel.leagues
let seasons = SportkalenderViewModel.seasons

File: android-native/app/src/main/java/de/sportkalender/nativeapp/SportkalenderViewModel.kt
package de.sportkalender.nativeapp

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import de.sportkalender.nativeapp.data.MatchRepository
import de.sportkalender.nativeapp.domain.resolveTvChannels
import de.sportkalender.nativeapp.model.LeagueOption
import de.sportkalender.nativeapp.model.OpenLigaMatch
import de.sportkalender.nativeapp.model.SeasonOption
import de.sportkalender.nativeapp.model.TvChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

data class MatchItemUi(
    val kickoff: String,
    val home: String,
    val away: String,
    val score: String,
    val tvChannels: List<TvChannel>
)

data class DayBlockUi(
    val dayLabel: String,
    val dateLabel: String,
    val isToday: Boolean,
    val matches: List<MatchItemUi>
)

data class UiState(
    val isLoading: Boolean = false,
    val error: String? = null,
    val selectedLeague: LeagueOption = LEAGUES.first(),
    val selectedSeason: SeasonOption = SEASONS.first(),
    val weekOffset: Int = 0,
    val weekLabel: String = "",
    val dayBlocks: List<DayBlockUi> = emptyList()
)

val LEAGUES = listOf(
    LeagueOption("bl1", "Bundesliga 1"),
    LeagueOption("bl2", "Bundesliga 2"),
    LeagueOption("bl3", "Bundesliga 3"),
    LeagueOption("dfb", "DFB-Pokal"),
    LeagueOption("ucl", "Champions League")
)

val SEASONS = listOf(
    SeasonOption("2025", "2025/26"),
    SeasonOption("2024", "2024/25"),
    SeasonOption("2023", "2023/24")
)

class SportkalenderViewModel(
    private val repository: MatchRepository = MatchRepository()
) : ViewModel() {

    private val _uiState = MutableStateFlow(UiState())
    val uiState: StateFlow<UiState> = _uiState.asStateFlow()

    private var allMatches: List<OpenLigaMatch> = emptyList()

    init {
        loadMatches()
    }

    fun onLeagueChanged(league: LeagueOption) {
        _uiState.value = _uiState.value.copy(selectedLeague = league, weekOffset = 0)
        loadMatches()
    }

    fun onSeasonChanged(season: SeasonOption) {
        _uiState.value = _uiState.value.copy(selectedSeason = season, weekOffset = 0)
        loadMatches()
    }

    fun previousWeek() {
        val newOffset = _uiState.value.weekOffset - 1
        _uiState.value = _uiState.value.copy(weekOffset = newOffset)
        recalcWeek()
    }

    fun nextWeek() {
        val newOffset = _uiState.value.weekOffset + 1
        _uiState.value = _uiState.value.copy(weekOffset = newOffset)
        recalcWeek()
    }

    fun goToCurrentWeek() {
        _uiState.value = _uiState.value.copy(weekOffset = 0)
        recalcWeek()
    }

    private fun loadMatches() {
        val league = _uiState.value.selectedLeague.id
        val season = _uiState.value.selectedSeason.id
        _uiState.value = _uiState.value.copy(isLoading = true, error = null)

        viewModelScope.launch(Dispatchers.IO) {
            runCatching { repository.loadMatches(league, season) }
                .onSuccess { matches ->
                    allMatches = matches
                    _uiState.value = _uiState.value.copy(isLoading = false, error = null)
                    recalcWeek()
                }
                .onFailure { throwable ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = throwable.message ?: "Unbekannter Fehler"
                    )
                }
        }
    }

    private fun recalcWeek() {
        val state = _uiState.value
        val monday = startOfCurrentWeek().plusDays((state.weekOffset * 7).toLong())
        val sunday = monday.plusDays(6)
        val dateFmt = DateTimeFormatter.ofPattern("dd.MM.yyyy", Locale.GERMANY)

        val blocks = (0L..6L).map { dayShift ->
            val date = monday.plusDays(dayShift)
            val dayMatches = allMatches
                .filter { match ->
                    val local = match.localKickoff()?.toLocalDate()
                    local == date
                }
                .sortedBy { it.localKickoff() }
                .map { match ->
                    val kickoff = match.localKickoff()
                    val score = finalScore(match) ?: "vs"
                    MatchItemUi(
                        kickoff = kickoff?.format(DateTimeFormatter.ofPattern("HH:mm")) ?: "--:--",
                        home = match.team1?.teamName ?: "Unbekannt",
                        away = match.team2?.teamName ?: "Unbekannt",
                        score = score,
                        tvChannels = resolveTvChannels(state.selectedLeague.id, kickoff)
                    )
                }

            DayBlockUi(
                dayLabel = date.dayOfWeek.getDisplayName(java.time.format.TextStyle.FULL, Locale.GERMAN),
                dateLabel = date.format(dateFmt),
                isToday = date == LocalDate.now(ZoneId.of("Europe/Berlin")),
                matches = dayMatches
            )
        }

        _uiState.value = state.copy(
            weekLabel = "${monday.format(dateFmt)} - ${sunday.format(dateFmt)}",
            dayBlocks = blocks
        )
    }

    private fun startOfCurrentWeek(): LocalDate {
        val today = LocalDate.now(ZoneId.of("Europe/Berlin"))
        return today.minusDays((today.dayOfWeek.value - 1).toLong())
    }

    private fun finalScore(match: OpenLigaMatch): String? {
        val final = match.matchResults
            ?.sortedByDescending { it.resultTypeId ?: 0 }
            ?.firstOrNull { it.pointsTeam1 != null && it.pointsTeam2 != null }
        return if (final != null) "${final.pointsTeam1}:${final.pointsTeam2}" else null
    }
}
